chain-tx-enclave-next/tx-query-next/app-runner/src/zmq_helper.rs

```diff
@@ -0,0 +1,189 @@
+use std::{
+    convert::TryInto,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+    vec::IntoIter,
+};
+
+use async_zmq::{request, Message, MultipartIter, Request};
+use futures::ready;
+use tokio::{
+    io::{self, AsyncRead, AsyncWrite},
+    sync::Mutex,
+};
+
+/// Temporary bridging between the old code and EDP. In the long term, it may go away completely.
+///
+/// 1. Decryption requests may include the sealed payload in them (so no need to request them via zmq)
+/// 2. Encryption requests may be done via a direct attested secure channel (instead of this passing of sealed requests)
+pub struct ZmqHelper {
+    inner: Mutex<Inner>,
+}
+
+/// Inner struct to hold values in a `Mutex`
+///
+/// `Mutex` is there to enforce the synchronous order as zmq isn't thread safe
+struct Inner {
+    /// ZeroMQ socket
+    socket: Request<IntoIter<Message>, Message>,
+    /// State of ZeroMQ socket
+    state: SocketState,
+}
+
+impl ZmqHelper {
+    pub fn new(endpoint: &str) -> io::Result<Self> {
+        let socket = request(endpoint)
+            .map_err(into_io_err)?
+            .connect()
+            .map_err(into_io_err)?;
+
+        Ok(ZmqHelper {
+            inner: Mutex::new(Inner {
+                socket,
+                state: SocketState::Ready,
+            }),
+        })
+    }
+}
+
+/// State of ZeroMQ socket
+#[derive(Debug)]
+enum SocketState {
+    /// Denotes that a new request can be sent on ZeroMQ socket
+    Ready,
+    /// Denotes that a request is sent to ZeroMQ socket but response is not received yet
+    RequestSent,
+    /// Denotes that response to a previously sent request is received
+    ResponseReceived(Message),
+}
+
+impl SocketState {
+    fn is_ready(&self) -> bool {
+        matches!(self, SocketState::Ready)
+    }
+
+    fn is_request_sent(&self) -> bool {
+        matches!(self, SocketState::RequestSent)
+    }
+
+    fn is_response_received(&self) -> bool {
+        matches!(self, SocketState::ResponseReceived(_))
+    }
+
+    fn unwrap(self) -> Message {
+        match self {
+            SocketState::ResponseReceived(message) => message,
+            _ => panic!("Cannot fetch response. Haven't received it yet."),
+        }
+    }
+}
+
+impl AsyncRead for ZmqHelper {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context,
+        buf: &mut [u8],
+    ) -> Poll<io::Result<usize>> {
+        let mut inner_future = Box::pin(self.inner.lock());
+        let mut inner = ready!(inner_future.as_mut().poll(cx));
+
+        if inner.state.is_ready() {
```

Comment:
why have these if-elses with `unreachable!` branch if one could just pattern match on `inner.state` directly ?

Reply:
Pattern matching on `inner.state` takes reference to the state and we cannot modify state if we have a reference to it. This is a workaround for that.

Reply:
Tried rewriting it as:
```
        let mut wakeup = false;
        let result = match inner_state {
            SocketState::Ready => {
                unreachable!("Unable to receive message. Previous request wasn't sent")
            },
            SocketState::RequestSent => {
                let response = {
                        let mut future = Box::pin(inner.socket.recv());
                        ready!(future.as_mut().poll(cx))
                    };
                match response {
                    Ok(multipart) => {
                        // We don't expect multipart messages from ZeroMQ. Assuming only one message
                        assert_eq!(
                            1,
                            multipart.len(),
                            "Received multiple messages in ZeroMQ response"
                        );

                        let response = multipart.into_iter().next().unwrap();
                        let response_len: u32 = response
                            .len()
                            .try_into()
                            .expect("Message length exceeds `u32` bounds");
                        buf.copy_from_slice(&response_len.to_le_bytes());

                        inner.state = SocketState::ResponseReceived(response);

                        Poll::Ready(Ok(std::mem::size_of::<u32>()))
                    }
                    Err(e) => Poll::Ready(Err(into_io_err(e))),
                }
            }
            SocketState::ResponseReceived(msg) => {
                buf.copy_from_slice(&msg);
                wakeup = true;

                Poll::Ready(Ok(msg.len()))
            }
        };
        if wakeup {
            inner.wake_next();
        }
        result
```
it seems to compile and is simpler?

Reply:
Both are pretty much the same things.

Reply:
I used the functions approach because that is the idiomatic way for doing the same thing for `Option`. [Example](https://github.com/crypto-com/chain/blob/a05b1cd101e3cab55a010fe86c9814bf116d903e/chain-abci/src/main.rs#L49).

Reply:
It's the same -- it's just a suggestion to avoid boilerplate, as it's perhaps 50 lines shorter with pattern match (no need for that `impl Socket` stuff etc.) and contains fewer foot guns for future code modifications (it's an exhaustive match, so no ```else {
             unreachable!()
        ``` } catch all)

Reply:
Ok. I’ll change.

Reply:
Changed.